### PR TITLE
Add event handling and communication for signature help.

### PR DESCRIPTION
**Pull request by steveluc**
Set up for method tips by adding mechanism to request signature help asynchronously.

### CODE REVIEW

File: TypeScript.py [hunk 1]
```diff
@@ -107,6 +107,10 @@ def extractLineOffset(lineOffset):
     offset = lineOffset.offset - 1
     return (line, offset)
 
+def span_to_region(span, view, extendAmt):
+    (startl, starto) = extractLineOffset(span.start)
+    (endl, endo) = extractLineOffset(span.end)
```
Comment (DanielRosenwasser) [inline TypeScript.py]: Use `startLine`, `endOffset` etc.

File: TypeScript.py [hunk 2]
```diff
@@ -107,6 +107,10 @@ def extractLineOffset(lineOffset):
     offset = lineOffset.offset - 1
     return (line, offset)
 
+def span_to_region(span, view, extendAmt):
```
Comment (DanielRosenwasser) [inline TypeScript.py]: `extendAmount`

File: TypeScript.py [hunk 3]
```diff
@@ -397,8 +402,9 @@ def setFilePrefs(view):
     settings = view.settings()
     settings.set('use_tab_stops', False)
     settings.set('translate_tabs_to_spaces', True)
-    settings.add_on_change('tab_size',lambda: tab_size_changed(view))
-    settings.add_on_change('indent_size',lambda: tab_size_changed(view))
+    if (settings and (settings.add_on_change)):
```
Comment (DanielRosenwasser) [inline TypeScript.py]: Omit parens

File: TypeScript.py [hunk 4]
```diff
@@ -397,8 +402,9 @@ def setFilePrefs(view):
...
+        settings.add_on_change('tab_size',lambda: tab_size_changed(view))
+        settings.add_on_change('indent_size',lambda: tab_size_changed(view))
```
Comment (DanielRosenwasser) [inline TypeScript.py]: Space after comma

File: TypeScript.py [hunk 5]
```diff
@@ -439,6 +445,49 @@ def checkUpdateView(view):
         if cli.reloadRequired(view):
             reloadBuffer(view, clientInfo)
 
+# these three methods work with signature help
+def handle_signature_help(signature_help_response):
+    if signature_help_response.success:
```
Comment (DanielRosenwasser) [inline TypeScript.py]: I feel like if you negate the conditions and bail out (i.e. `not signature_help_response.success`), you end up with something less nested and more readable

Comment (DanielRosenwasser) [inline TypeScript.py]: Also we're very inconsistent about casing throughout, consider renaming the parameter sy least to match its locals.

File: TypeScript.py [hunk 6]
```diff
@@ -439,6 +445,49 @@ def checkUpdateView(view):
...
+                print("show method tip UI")
+
+# check trigger for signature help
+def check_trigger_key(key, view):
+    if cli.ST2():
+        return
+    print('check trigger key ' + key)
+    if (key == '(') or (key == ','):
+        print('yes trigger')
+        cli.service.asyncSignatureHelp(view.file_name(),
```
Comment (DanielRosenwasser) [inline TypeScript.py]: Space after commas.

File: TypeScript.py [hunk 7]
```diff
@@ -439,6 +445,49 @@ def checkUpdateView(view):
...
+                                       getLocationFromView(view),'', handle_signature_help)
```
Comment (DanielRosenwasser) [inline TypeScript.py]: Space after commas.

File: TypeScript.py [hunk 8]
```diff
@@ -439,6 +445,49 @@ def checkUpdateView(view):
...
+                # put up method tip UI
+                clientInfo.method_tip_active = True
+                print("show method tip UI")
+
+# check trigger for signature help
+def check_trigger_key(key, view):
+    if cli.ST2():
+        return
+    print('check trigger key ' + key)
+    if (key == '(') or (key == ','):
```
Comment (DanielRosenwasser) [inline TypeScript.py]: For the record in VS we also have trigger characters to support tagged templates. I can look into it tomorrow.

File: TypeScript.py (hunk 6 above)
Comment (DanielRosenwasser) [inline TypeScript.py]: Should these print statements still be here or are they still useful?

File: TypeScript.py [hunk 9]
```diff
@@ -731,7 +780,7 @@ def on_close(self, view):
                     # notify the server that the file is closed
                     cli.service.close(view.file_name())
         logger.view_debug(view, "exit on_close")
-
+
```
Comment (DanielRosenwasser) [inline TypeScript.py]: Remove

File: TypeScript.py [hunk 10]
```diff
@@ -439,6 +445,49 @@ def checkUpdateView(view):
...
+    if cli.ST2():
+        return
+    print('check trigger key ' + key)
+    if (key == '(') or (key == ','):
+        print('yes trigger')
+        cli.service.asyncSignatureHelp(view.file_name(),
+                                       getLocationFromView(view),'', handle_signature_help)
+
+# check whether cursor is outside the current applicable span for signature help                                   
+def check_outside_applicable_span():
```
Comment (DanielRosenwasser) [inline TypeScript.py]: This is misleading - I'd make this `isOutsideApplicableSpan` and make a `dismissSignatureHelp` that you call if it returns true

File: TypeScript.py [hunk 11]
```diff
@@ -439,6 +445,49 @@ def checkUpdateView(view):
...
+        return 
+    view = active_view()
+    if is_typescript(view):
+        clientInfo = cli.getOrAddFile(view.file_name())
+        if clientInfo.method_tip_active:
+            pos = view.sel()[0].begin()
+            applicable_region = view.get_regions("signatureApplicable")[0]
+            if applicable_region and ((not applicable_region.contains(pos)) or (pos == applicable_region.end())):
+                # dismiss method tip
+                print("dismiss method tip ")
```
Comment (zhengbli) [inline TypeScript.py]: might be cleaner to use logger.log.debug/warn/info instead of print for logging

Comment (DanielRosenwasser) [inline TypeScript.py]: :+1: Agreed, use standard logging facilities when possible

File: TypeScript.py (hunk 4 above)
Comment (billti) [inline TypeScript.py]: Is this related to signature help?  Or just some changes that got rolled in here?

**Comment by billti** [comment]
Thanks Steve!  I think I incorporated most of these ideas into my tooltips branch to implement signature help.  Is this pull request required for any other changes?

**Comment by steveluc** [comment]
Closing this.

File: TypeScript.py (hunk 4 above)
Comment (steveluc) [inline TypeScript.py]: These are un-related but I think subsumed by later changes.  They are for updating format configuration when it changes per file.
